### PR TITLE
(210) Add e2e tests for reports

### PR DESCRIPTION
**Pull request by libuk**
[Jira](https://dsdmoj.atlassian.net/browse/CAS2-210?atlOrigin=eyJpIjoiNDU0MDI0YzVkZWU0NGFkYWExNmJhN2E1YjYxOTllY2UiLCJwIjoiaiJ9)

# Context
Test that a user can download reports and that the spreadsheet contains the expected columns.

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Are any changes required to the e2e tests?
- [ ] If you've added a new route, have you added a new
  `auditEvent`? (see `server/routes/apply.ts` for examples)
- [ ] Are there environment variables or other infrastructure configuration which needs to be included in this release?
- [ ] Are there any data migrations required. Automatic or manual?
- [ ] Does this rely on changes being deployed to the CAS API?

## Post merge checklist

Once we've merged it will be auto-deployed to the dev environment.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-community-accommodation-tier-2-ui).

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

File: e2e-tests/steps/reports.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+export const confirmColumnNames = async (reportType: ReportType, path: string) => {
+  const workbook = new Excel.Workbook()
+
+  await workbook.xlsx.readFile(path).then(() => {
+    const sh = workbook.getWorksheet('Sheet0')
+
+    const headerCells: CellValue[] = []
+    sh.getRow(1).eachCell(cell => headerCells.push(cell.value))
+
+    expect(headerCells.length === reportMetaData[reportType].columnNames.length).toBe(true)
```
Comment (libuk) [inline e2e-tests/steps/reports.ts:66]: This assertion means that if the report is updated on the API side without an update on the FE side this test will fail. I was in two minds about this scenario blocking the FE pipeline, but let me know your thoughts.

File: e2e-tests/steps/reports.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+export const confirmColumnNames = async (reportType: ReportType, path: string) => {
+  const workbook = new Excel.Workbook()
+
+  await workbook.xlsx.readFile(path).then(() => {
+    const sh = workbook.getWorksheet('Sheet0')
+
+    const headerCells: CellValue[] = []
+    sh.getRow(1).eachCell(cell => headerCells.push(cell.value))
+
+    expect(headerCells.length === reportTypeMetaData[reportType].columnNames.length).toBe(true)
```
Comment (libuk) [inline e2e-tests/steps/reports.ts:66]: This assertion means that if the report is updated with a new column on the API without an update on the FE this test will fail. I was in two minds about this scenario blocking the FE pipeline, but let me know your thoughts.

Comment (patrickjfl) [reply e2e-tests/steps/reports.ts:61]: Is this in place to make sure we're not leaking data unexpectedly, or to make sure we keep our E2E tests up to date? If it's the latter then I'm not sure the best way to control that is with tight tests, at the expense of blocking the FE pipeline as you say. Not a major deal though.

File: e2e-tests/steps/reports.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    ],
+    callToAction: "Download 'Submitted applications' report",
+  },
+  applicationStatusUpdates: {
+    columnNames: ['eventId', 'applicationId', 'personCrn', 'personNoms', 'updatedBy', 'updatedAt', 'newStatus'],
+    callToAction: "Download 'Application status updates' report",
+  },
+  unsubmittedApplications: {
+    columnNames: ['applicationId', 'personCrn', 'personNoms', 'startedBy', 'startedAt'],
+    callToAction: "Download 'Un-submitted applications' report",
```
Comment (patrickjfl) [inline e2e-tests/steps/reports.ts:27]: I guess this is coming from the API, but 'un-submitted' applications feels weird content wise.

**Review by patrickjfl — APPROVED** [review]

**Comment by patrickjfl** [comment]
Approved but it would be good to test these locally - where can I find the local MI user credentials?

**Comment by libuk** [comment]
@patrickjfl 

> Is this in place to make sure we're not leaking data unexpectedly, or to make sure we keep our E2E tests up to date? If it's the latter then I'm not sure the best way to control that is with tight tests, at the expense of blocking the FE pipeline as you say. Not a major deal though.

It was for the latter, I'll remove as it's probably overkill 👍🏾 

> Approved but it would be good to test these locally - where can I find the local MI user credentials?

You can find the credentials in 1password

File: e2e-tests/steps/reports.ts (hunk 3 above)
Comment (libuk) [reply e2e-tests/steps/reports.ts:27]: This matches the content in our FE, so we can change this at any point. I'll run it past Chanel 👍🏾

**Comment by patrickjfl** [comment]
> @patrickjfl
> 
> > Is this in place to make sure we're not leaking data unexpectedly, or to make sure we keep our E2E tests up to date? If it's the latter then I'm not sure the best way to control that is with tight tests, at the expense of blocking the FE pipeline as you say. Not a major deal though.
> 
> It was for the latter, I'll remove as it's probably overkill 👍🏾
> 
> > Approved but it would be good to test these locally - where can I find the local MI user credentials?
> 
> You can find the credentials in 1password

Ah thanks, wasn't sure if the dev ones would work locally. The tests are passing locally for me 👍

File: e2e-tests/steps/reports.ts (hunk 3 above)
Comment (libuk) [reply e2e-tests/steps/reports.ts:27]: Leaving as is for now, but the conversation is here: https://mojdt.slack.com/archives/C04T6SC31HA/p1709137791628899

**Pull request merged**